Modules/Core/Common/include/itkMacro.h

```diff
@@ -343,6 +343,28 @@ namespace itk
   }                                                                       \
   ITK_MACROEND_NOOP_STATEMENT
 
+/** Define an object creation method throwing an exception if the object
+ * is not created through the object factory, for use in base classes that
+ * do not fully implement a backend. */
+#define itkFactoryOnlyNewMacro(x)  \
+  itkSimpleFactoryOnlyNewMacro(x); \
+  itkCreateAnotherMacro(x);        \
+  itkCloneMacro(x);                \
+  ITK_MACROEND_NOOP_STATEMENT
+
+#define itkSimpleFactoryOnlyNewMacro(x)                                                             \
+  static auto New()->Pointer                                                                        \
+  {                                                                                                 \
+    Pointer smartPtr = itk::ObjectFactory<x>::Create();                                             \
+    if (smartPtr == nullptr)                                                                        \
+    {                                                                                               \
+      itkSpecializedMessageExceptionMacro(ExceptionObject, "Object factory failed to instantiate"); \
```

Comment:
> I think it would be good to include the name of the class or factory used in this error message.

I agree in theory, but unfortunately we've documented unresolved compiler issues with `Self->GetNameOfClass()`, which would be required here for printing the class name (see #3008).

We can't print a factory name because the error results from the absence of a suitable factory.

`itkErrorMacro` by default returns the location of the error along with the error message, which should be suitable for error tracing.

> A similar condition happens in the ImageFileReader:

Looking at commit [a5b416c](https://github.com/InsightSoftwareConsortium/ITK/commit/a5b416c6b45f83d08617fdb4a409af718df510cc) which introduced the changes, it looks like the chief purpose of that error message was to indicate that the object factory is involved, which we do here with the standard message from `itkFactoryOnlyNewMacro`. 

I wonder a bit about whether we really want to include URLs directly to a wiki (I'm not sure whether [https://www.itk.org/Wiki/ITK](https://www.itk.org/Wiki/ITK) is still maintained?), but especially in the case of defining a macro it would be best to avoid overspecifying the exception message.

Reply:
> I wonder a bit about whether we really want to include URLs directly to a wiki (I'm not sure whether https://www.itk.org/Wiki/ITK is still maintained?), but especially in the case of defining a macro it would be best to avoid overspecifying the exception message.

Yes, I think the URL is a little dated. And is not the best way to do things.

If smartPtr is nullptr we don't have an object to call `GetNameOfClassOn`. The requested class is what is needed which may be `typeid(x).name()`.

Reply:
> The requested class is what is needed which may be typeid(x).name().

This is workable, it looks like we do this elsewhere for logging in ITK too. I will add this.